### PR TITLE
Use @ to separate architecture from subrepo names

### PR DESCRIPTION
**Pull request by Tatskaari**
This allows us to pass architecture for subrepo targets on the command line and makes it a little easier to reason about. I've seen a few issues when testing where it locks up or fails unexpectedly. I think this is probably too risky to go in with the virutalised FS stuff so will hold off for now.

### CODE REVIEW

File: src/parse/parse_step.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func checkSubrepo(state *core.BuildState, label, dependent core.BuildLabel, mode
 	}
 
 	// Try parsing the package in the host repo first.
-	s, err := parseSubrepoPackage(state, sl.PackageName, "", label, mode)
+	s, err := parseSubrepoPackage(state, sl.PackageName, sl.Subrepo, label, mode)
```
Comment (Tatskaari) [inline src/parse/parse_step.go:112]: sl.Subrepo is now set to the arch part i.e. for `///go@darwin_amd64//build_defs:go`, sl is `BuildLabel{PackageName: "go", Subrepo: "darwin_amd64"}`

**Review by samwestmoreland — APPROVED** [review]
Nice!

File: src/parse/asp/interpreter.go [hunk 2]
```diff
@@ -306,18 +306,41 @@ func (s *scope) parseAnnotatedLabelInPackage(label string, pkg *core.Package) co
...
-			subrepo = pkg.SubrepoName
-		} else if arch := cli.HostArch(); strings.Contains(subrepo, "_"+arch.String()) {
-			subrepo = strings.TrimSuffix(subrepo, "_"+arch.String())
-		} else if subrepo == arch.String() {
+		arch := cli.HostArch()
+
+		subrepo, subrepoArch := core.SplitSubrepoArch(subrepo)
+
+		// Strip out the host arch as that's the default
+		if subrepo == arch.String() {
```
[the remark this answers is not in the document]
Comment (Tatskaari) [reply src/parse/asp/interpreter.go:314]: I think a v18 change will be to make `///darwin_amd64` illegal, requiring that you use the `@darwin_amd64` syntax for architectures, and the `///subrepo` syntax for subrepos. This is a breaking change though :(

File: src/core/build_label.go [hunk 3]
```diff
@@ -415,37 +435,47 @@ func (label BuildLabel) PackageDir() string {
...
 // SubrepoLabel returns a build label corresponding to the subrepo part of this build label.
-func (label BuildLabel) SubrepoLabel(state *BuildState, dependentSubrepo string) BuildLabel {
-	arch := ""
-	if dependentSubrepo != "" {
-		a := state.Graph.Subrepo(dependentSubrepo).Arch
-		if a != state.Arch {
-			arch = a.String()
-		}
+func (label BuildLabel) SubrepoLabel(state *BuildState) BuildLabel {
+	pluginName, arch := SplitSubrepoArch(label.Subrepo)
```
[the remark this answers is not in the document]
Comment (Tatskaari) [reply src/core/build_label.go]: Yup already changed :)

**Pull request merged**